### PR TITLE
Better rendering of bech32 error messages

### PR DESCRIPTION
**Pull request by Anviking**
Like #1619, but with `master` as base branch 🤦‍♂️

### CODE REVIEW

**Review by KtorZ — APPROVED** [review]
bors r+

**Comment by Anviking** [comment]
```
test/integration/Test/Integration/Byron/Scenario/CLI/Transactions.hs:563:56:
  1) CLI Specifications, BYRON_TXS_CLI, CLI_TRANS_ESTIMATE_01/02 - 2 recipient(s)
       uncaught exception: RequestException
       DecodeFailure "Error in $: parsing [] failed, expected Array, but encountered Object: Response {responseStatus = Status {statusCode = 404, statusMessage = \"Not Found\"}, responseVersion = HTTP/1.1, responseHeaders = [(\"Transfer-Encoding\",\"chunked\"),(\"Date\",\"Wed, 06 May 2020 08:11:29 GMT\"),(\"Server\",\"Warp/3.3.5\")], responseBody = \"{\\\"code\\\":\\\"no_such_wallet\\\",\\\"message\\\":\\\"I couldn't find a wallet with the given id: da15d905f705267a54b8f1fc74251fb3fc720cc5\\\"}\", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}"

  To rerun use: --match "/CLI Specifications/BYRON_TXS_CLI/CLI_TRANS_ESTIMATE_01/02 - 2 recipient(s)/"

  test/integration/Test/Integration/Byron/Scenario/CLI/Transactions.hs:713:31:
  2) CLI Specifications, BYRON_TXS_CLI, CLI_TRANS_ESTIMATE_04 - Error shown when ErrInputsDepleted encountered
       uncaught exception: RequestException
       DecodeFailure "Error in $: parsing [] failed, expected Array, but encountered Object: Response {responseStatus = Status {statusCode = 404, statusMessage = \"Not Found\"}, responseVersion = HTTP/1.1, responseHeaders = [(\"Transfer-Encoding\",\"chunked\"),(\"Date\",\"Wed, 06 May 2020 08:11:32 GMT\"),(\"Server\",\"Warp/3.3.5\")], responseBody = \"{\\\"code\\\":\\\"no_such_wallet\\\",\\\"message\\\":\\\"I couldn't find a wallet with the given id: da15d905f705267a54b8f1fc74251fb3fc720cc5\\\"}\", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}"

  To rerun use: --match "/CLI Specifications/BYRON_TXS_CLI/CLI_TRANS_ESTIMATE_04 - Error shown when ErrInputsDepleted encountered/"

  test/integration/Test/Integration/Byron/Scenario/CLI/Transactions.hs:736:31:
  3) CLI Specifications, BYRON_TXS_CLI, CLI_TRANS_ESTIMATE_04 - Can't cover fee
       uncaught exception: RequestException
       DecodeFailure "Error in $: parsing [] failed, expected Array, but encountered Object: Response {responseStatus = Status {statusCode = 404, statusMessage = \"Not Found\"}, responseVersion = HTTP/1.1, responseHeaders = [(\"Transfer-Encoding\",\"chunked\"),(\"Date\",\"Wed, 06 May 2020 08:11:38 GMT\"),(\"Server\",\"Warp/3.3.5\")], responseBody = \"{\\\"code\\\":\\\"no_such_wallet\\\",\\\"message\\\":\\\"I couldn't find a wallet with the given id: da15d905f705267a54b8f1fc74251fb3fc720cc5\\\"}\", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}"

  To rerun use: --match "/CLI Specifications/BYRON_TXS_CLI/CLI_TRANS_ESTIMATE_04 - Can't cover fee/"
```

bors r+

**Comment by rvl** [comment]
Those test failures look like a problem with flaky test setup. It says the wallet doesn't exist.
The failures are on macOS by the way.

File: lib/cli/src/Cardano/CLI.hs [hunk 1]
```diff
@@ -501,6 +504,30 @@ decodeKey = \case
...
+                    <> markCharsRedAtIndices (map unCharPos ixs) (T.unpack t)
+        invalidCharsMsg = "Invalid character(s) in string"
+        unCharPos (CharPosition x) = x
+
+markCharsRedAtIndices :: Integral i => [i] -> String -> String
+markCharsRedAtIndices ixs txt = go 0 ixs txt ++ def
+  where
+    go _c [] [] = mempty
+    go c (i:is) (s:ss)
+        | c == i    = red ++ s:def ++ go (c + 1) is ss
```
Comment (rvl) [inline lib/cli/src/Cardano/CLI.hs:526]: This will output 2 ansi escape sequences for each invalid character.

Comment (Anviking) [reply lib/cli/src/Cardano/CLI.hs:526]: Yes, but it still works, and multiple invalid characters should be rare.

**Comment by Anviking** [comment]
*did rebasing help?*
bors r+

File: lib/cli/src/Cardano/CLI.hs [hunk 2]
```diff
@@ -504,6 +507,30 @@ decodeKey = \case
...
+            StringToDecodeHasMixedCase -> "string has mixed case"
+            StringToDecodeMissingSeparatorChar -> "string has no separator char"
+            StringToDecodeContainsInvalidChars [] -> invalidCharsMsg
+            StringToDecodeContainsInvalidChars ixs -> invalidCharsMsg <> ":\n"
+                    <> markCharsRedAtIndices (map unCharPos ixs) (T.unpack t)
+        invalidCharsMsg = "Invalid character(s) in string"
+        unCharPos (CharPosition x) = x
+
+markCharsRedAtIndices :: Integral i => [i] -> String -> String
+markCharsRedAtIndices ixs txt = go 0 ixs txt ++ def
```
Comment (Anviking) [inline lib/cli/src/Cardano/CLI.hs:522]: Should perhaps sort the `ixs` here 🤔

Comment (Anviking) [reply lib/cli/src/Cardano/CLI.hs:522]: Think it's off, will follow up with property test and fix

**Pull request merged**